File: tests/test_web.ts
/** Tests for the read-only web dashboard: payloads, SVG charts and routes. */
import { strict as assert } from 'node:assert';
import { test } from 'node:test';

import { Database } from '../hera/database';
import { MarketService } from '../hera/services/market';
import { Random } from '../hera/market/random';
import { config } from '../hera/config';
import {
  chartDataUri,
  fallbackChart,
  renderComparisonChart,
  renderPriceChart,
  renderSparkline,
} from '../dashboard/charts';
import { DashboardService } from '../dashboard/service';
import { Dashboard } from '../dashboard/server';
import { testMarketConfig, GUILD } from './harness';

function advance(market: MarketService, count: number): void {
  for (let i = 0; i < count; i += 1) market.tick(GUILD);
}

interface Fixture {
  db: Database;
  market: MarketService;
  service: DashboardService;
  close(): void;
}

function fixture(): Fixture {
  const db = new Database(':memory:');
  const market = new MarketService(db, testMarketConfig(), new Random(1234));
  advance(market, 40);
  return {
    db,
    market,
    service: new DashboardService(market, GUILD),
    close: () => db.close(),
  };
}

// ------------------------------------------------------------------ the payload

test('the market payload lists every company', async () => {
  const f = fixture();
  const payload = await f.service.marketPayload();
  assert.equal(payload.stocks.length, 20);
  assert.equal(payload.totals.listings, 20);
  assert.ok(payload.tick > 0);
  f.close();
});

test('the market payload has the documented shape', async () => {
  const f = fixture();
  const payload = await f.service.marketPayload();
  for (const key of ['generated_at', 'tick', 'regime', 'index', 'totals', 'sectors', 'stocks', 'news']) {
    assert.ok(key in payload, `missing ${key}`);
  }
  assert.deepEqual(Object.keys(payload.index).sort(), ['change', 'change_pct', 'value']);
  f.close();
});

test('stock rows expose the tracking fields', async () => {
  const f = fixture();
  const row = (await f.service.marketPayload()).stocks[0];
  for (const key of [
    'symbol',
    'name',
    'sector',
    'price',
    'change_pct',
    'day_high',
    'day_low',
    'market_cap',
    'dividend_yield',
    'halted',
    'sparkline',
  ]) {
    assert.ok(key in row, `missing ${key}`);
  }
  assert.ok(row.sparkline.startsWith('data:image/svg+xml;base64,'));
  f.close();
});

test('change_pct matches price and previous_close', async () => {
  const f = fixture();
  const payload = await f.service.marketPayload();
  for (const row of payload.stocks) {
    const expected = row.previous_close ? ((row.price - row.previous_close) / row.previous_close) * 100 : 0;
    assert.ok(Math.abs(row.change_pct - expected) < 1e-9, `${row.symbol} change_pct`);
  }
  f.close();
});

test('sectors cover every stock', async () => {
  const f = fixture();
  const payload = await f.service.marketPayload();
  const counted = payload.sectors.reduce((sum, sector) => sum + sector.count, 0);
  assert.equal(counted, payload.stocks.length);
  f.close();
});

test('totals breadth adds up', async () => {
  const f = fixture();
  const { totals } = await f.service.marketPayload();
  assert.equal(totals.gainers + totals.losers + totals.unchanged, totals.listings);
  f.close();
});

// -------------------------------------------------------------------- the charts

test('the price chart is well-formed SVG with the series', () => {
  const history: [number, number][] = Array.from({ length: 30 }, (_, i) => [i, 100 + i]);
  const svg = renderPriceChart('TEST', 'Test Co', history);
  assert.ok(svg.startsWith('<svg'));
  assert.ok(svg.trimEnd().endsWith('</svg>'));
  assert.ok(svg.includes('TEST') && svg.includes('Test Co'));
  assert.equal(svg.split('<polyline').length - 1, 1);
});

test('the price chart needs at least two points', () => {
  assert.equal(renderPriceChart('TEST', 'Test Co', [[1, 10]]), '');
  assert.equal(renderPriceChart('TEST', 'Test Co', []), '');
});

test('the sparkline requires two points', () => {
  assert.equal(renderSparkline([]), '');
  assert.equal(renderSparkline([5]), '');
  const svg = renderSparkline([1, 2, 3]);
  assert.ok(svg.startsWith('<svg') && svg.includes('</svg>'));
});

test('the comparison chart normalises and handles empty input', () => {
  assert.equal(renderComparisonChart({}), '');
  assert.equal(renderComparisonChart({ A: [1] }), '');
  const svg = renderComparisonChart({ A: [100, 110], B: [50, 45] });
  assert.equal(svg.split('<polyline').length - 1, 2);
  assert.ok(svg.includes('A') && svg.includes('B'));
});

test('the chart data URI round-trips', () => {
  const uri = chartDataUri('<svg></svg>');
  assert.ok(uri.startsWith('data:image/svg+xml;base64,'));
  assert.equal(Buffer.from(uri.split(',')[1], 'base64').toString('utf-8'), '<svg></svg>');
  assert.equal(chartDataUri(''), '');
});

test('the fallback chart is a valid image', () => {
  const svg = fallbackChart();
  assert.ok(svg.startsWith('<svg') && svg.trimEnd().endsWith('</svg>'));
});

test('stockChartSvg renders a known symbol', () => {
  const f = fixture();
  const svg = f.service.stockChartSvg('NOVA');
  assert.ok(svg);
  assert.ok(svg.includes('NOVA') && svg.startsWith('<svg'));
  assert.ok(f.service.stockChartSvg('nova'));
  assert.equal(f.service.stockChartSvg('ZZZZ'), null);
  f.close();
});

test('indexChartSvg renders the index', () => {
  const f = fixture();
  const svg = f.service.indexChartSvg();
  assert.ok(svg && svg.includes('Hera Exchange Index'));
  f.close();
});

test('rendered SVG has no unfilled placeholders', () => {
  // Guards against a template literal silently slipping into the markup.
  const svg = renderPriceChart('TEST', 'Test Co', Array.from({ length: 10 }, (_, i) => [i, 10 + i]));
  assert.equal(/\{[a-z_]+\}/.test(svg), false);
});

// --------------------------------------------------------------------- the routes

interface Response {
  status: number;
  contentType: string | null;
  body: string;
}

async function get(port: number, path: string): Promise<Response> {
  const response = await fetch(`http://127.0.0.1:${port}${path}`, {
    headers: { accept: '*/*' },
  });
  return {
    status: response.status,
    contentType: response.headers.get('content-type'),
    body: await response.text(),
  };
}

/** Start a live dashboard on a port the OS picks, and return its details. */
async function withServer<T>(body: (port: number, dashboard: Dashboard) => Promise<T>): Promise<T> {
  const { mkdtempSync, rmSync } = await import('node:fs');
  const { tmpdir } = await import('node:os');
  const { join } = await import('node:path');

  const directory = mkdtempSync(join(tmpdir(), 'hera-dashboard-'));
  const path = join(directory, 'test.db');
  const previous = { dbPath: config.dbPath, guildId: config.guildId };

  // Seed the file the dashboard will open, then point the shared config at it.
  const seed = new Database(path);
  const market = new MarketService(seed, testMarketConfig(), new Random(1));
  for (let i = 0; i < 40; i += 1) market.tick(GUILD);
  seed.close();
  config.dbPath = path;
  config.guildId = GUILD;

  const dashboard = new Dashboard();
  await dashboard.start('127.0.0.1', 0);
  const address = (dashboard as unknown as { server: { address(): { port: number } } }).server.address();
  try {
    return await body(address.port, dashboard);
  } finally {
    await dashboard.stop();
    config.dbPath = previous.dbPath;
    config.guildId = previous.guildId;
    rmSync(directory, { recursive: true, force: true });
  }
}

test('the overview and stocks pages are served', async () => {
  await withServer(async (port) => {
    for (const path of ['/', '/stocks']) {
      const response = await get(port, path);
      assert.equal(response.status, 200);
      assert.ok(response.body.includes('<!DOCTYPE html>'));
      assert.ok(response.body.includes('/static/app.js'));
    }
  });
});

test('static assets are served', async () => {
  await withServer(async (port) => {
    for (const path of ['/static/style.css', '/static/app.js']) {
      const response = await get(port, path);
      assert.equal(response.status, 200);
      assert.ok(response.body.trim().length > 0);
    }
  });
});

test('the market JSON endpoint responds', async () => {
  await withServer(async (port) => {
    const response = await get(port, '/api/market');
    assert.equal(response.status, 200);
    const payload = JSON.parse(response.body);
    assert.ok(payload.stocks.length > 0);
    assert.ok(payload.index.value > 0);
  });
});

test('the stocks JSON endpoint responds', async () => {
  await withServer(async (port) => {
    const response = await get(port, '/api/stocks');
    assert.equal(response.status, 200);
    assert.equal(JSON.parse(response.body).stocks.length, 20);
  });
});

test('the chart endpoint returns SVG', async () => {
  await withServer(async (port) => {
    const response = await get(port, '/api/charts/nova.svg');
    assert.equal(response.status, 200);
    assert.equal(response.contentType, 'image/svg+xml');
    assert.ok(response.body.startsWith('<svg'));
    assert.ok(response.body.includes('NOVA'));
  });
});

test('the index chart endpoint returns SVG', async () => {
  await withServer(async (port) => {
    const response = await get(port, '/api/charts/index.svg');
    assert.equal(response.status, 200);
    assert.equal(response.contentType, 'image/svg+xml');
    assert.ok(response.body.startsWith('<svg'));
  });
});

test('the chart endpoint 404s on an unknown symbol', async () => {
  await withServer(async (port) => {
    assert.equal((await get(port, '/api/charts/zzzz.svg')).status, 404);
  });
});

test('the health endpoint responds', async () => {
  await withServer(async (port) => {
    const response = await get(port, '/healthz');
    assert.equal(response.status, 200);
    const payload = JSON.parse(response.body);
    assert.equal(payload.status, 'ok');
    assert.ok(payload.tick > 0);
  });
});

test('the dashboard is read-only', async () => {
  await withServer(async (port) => {
    const before = await get(port, '/api/market');
    for (const path of [
      '/',
      '/stocks',
      '/api/market',
      '/api/stocks',
      '/api/charts/nova.svg',
      '/api/charts/index.svg',
      '/healthz',
    ]) {
      await get(port, path);
    }
    const after = await get(port, '/api/market');
    const beforePayload = JSON.parse(before.body);
    const afterPayload = JSON.parse(after.body);
    assert.equal(afterPayload.tick, beforePayload.tick);
    assert.deepEqual(
      afterPayload.stocks.map((row: { price: number }) => row.price),
      beforePayload.stocks.map((row: { price: number }) => row.price),
    );
  });
});

test('the dashboard refuses to start without a guild', () => {
  const previous = config.guildId;
  config.guildId = null;
  try {
    assert.throws(() => new Dashboard(), /guild/i);
  } finally {
    config.guildId = previous;
  }
});

test('the dashboard can serve an injected market without owning the database', async () => {
  const db = new Database(':memory:');
  const market = new MarketService(db, testMarketConfig(), new Random(99));
  advance(market, 25);
  const dashboard = new Dashboard({ guildId: GUILD, db, market });
  await dashboard.start('127.0.0.1', 0);
  const address = (dashboard as unknown as { server: { address(): { port: number } } }).server.address();
  try {
    const response = await get(address.port, '/api/market');
    assert.equal(response.status, 200);
    const payload = JSON.parse(response.body);
    assert.equal(payload.stocks.length, 20);
    assert.equal(payload.tick, 25);
  } finally {
    // stop() must leave a borrowed database open for its owner.
    await dashboard.stop();
  }
  // The bot relies on this: the handle is still usable, and it closes it itself
  // during shutdown.
  const stillOpen = new Dashboard({ guildId: GUILD, db, market });
  await stillOpen.start('127.0.0.1', 0);
  await stillOpen.stop();
  db.close();
});

test('a dashboard opened standalone closes the database it owns', async () => {
  const previous = config.dbPath;
  const { mkdtempSync, rmSync } = await import('node:fs');
  const { tmpdir } = await import('node:os');
  const { join } = await import('node:path');
  const directory = mkdtempSync(join(tmpdir(), 'hera-dash-'));
  const path = join(directory, 'hera.db');
  const seed = new Database(path);
  const market = new MarketService(seed, testMarketConfig(), new Random(3));
  advance(market, 5);
  seed.close();
  config.dbPath = path;
  try {
    const dashboard = new Dashboard({ guildId: GUILD });
    await dashboard.start('127.0.0.1', 0);
    await dashboard.stop();
    // It opened (and closed) its own connection, so reopening the file works.
    const reopened = new Database(path);
    const count = reopened.fetchval<number>('SELECT COUNT(*) FROM companies');
    assert.equal(count, 20);
    reopened.close();
  } finally {
    config.dbPath = previous;
    rmSync(directory, { recursive: true, force: true });
  }
});

File: tests/test_web.py
"""Tests for the read-only web dashboard: service payloads, SVG charts and routes.

These run against a real temporary SQLite database and a real aiohttp server, so
the JSON and the images a browser would receive are exercised end to end.
"""

from __future__ import annotations

import re

import pytest
from aiohttp.test_utils import TestClient, TestServer

from hera.web.charts import (
    chart_data_uri,
    render_comparison_chart,
    render_price_chart,
    render_sparkline,
)
from hera.web.server import Dashboard
from hera.web.service import DashboardService, fallback_chart

GUILD = 424242


async def _advance(market, count: int) -> None:
    for _ in range(count):
        await market.tick(GUILD)


@pytest.fixture
async def dashboard_service(market):
    await _advance(market, 40)
    return DashboardService(market, guild_id=GUILD)


@pytest.fixture
async def client(market):
    await _advance(market, 40)
    dashboard = Dashboard(market, guild_id=GUILD)
    async with TestClient(TestServer(dashboard.app)) as test_client:
        yield test_client


# ------------------------------------------------------------------ the payload


async def test_market_payload_lists_every_company(dashboard_service, company_count):
    payload = await dashboard_service.market_payload()
    assert len(payload["stocks"]) == company_count
    assert payload["totals"]["listings"] == company_count
    assert payload["tick"] > 0


async def test_market_payload_shape(dashboard_service):
    payload = await dashboard_service.market_payload()
    for key in ("generated_at", "tick", "regime", "index", "totals", "sectors", "stocks", "news"):
        assert key in payload
    assert set(payload["index"]) == {"value", "change", "change_pct"}


async def test_stock_rows_expose_the_tracking_fields(dashboard_service):
    payload = await dashboard_service.market_payload()
    row = payload["stocks"][0]
    for key in (
        "symbol",
        "name",
        "sector",
        "price",
        "change_pct",
        "day_high",
        "day_low",
        "market_cap",
        "dividend_yield",
        "halted",
        "sparkline",
    ):
        assert key in row
    assert row["sparkline"].startswith("data:image/svg+xml;base64,")


async def test_change_pct_matches_price_and_previous_close(dashboard_service):
    payload = await dashboard_service.market_payload()
    for row in payload["stocks"]:
        expected = (
            (row["price"] - row["previous_close"]) / row["previous_close"] * 100
            if row["previous_close"]
            else 0.0
        )
        assert row["change_pct"] == pytest.approx(expected)


async def test_sectors_cover_every_stock(dashboard_service):
    payload = await dashboard_service.market_payload()
    counted = sum(sector["count"] for sector in payload["sectors"])
    assert counted == len(payload["stocks"])


async def test_totals_breadth_adds_up(dashboard_service):
    payload = await dashboard_service.market_payload()
    totals = payload["totals"]
    assert totals["gainers"] + totals["losers"] + totals["unchanged"] == totals["listings"]


# -------------------------------------------------------------------- the charts


def test_price_chart_is_wellformed_svg_with_the_series():
    history = [(i, 100.0 + i) for i in range(30)]
    svg = render_price_chart("TEST", "Test Co", history)
    assert svg.startswith("<svg")
    assert svg.rstrip().endswith("</svg>")
    assert "TEST" in svg and "Test Co" in svg
    # One tick label per sampled x position, and at least one gridline.
    assert svg.count("<polyline") == 1


def test_price_chart_needs_at_least_two_points():
    assert render_price_chart("TEST", "Test Co", [(1, 10.0)]) == ""
    assert render_price_chart("TEST", "Test Co", []) == ""


def test_sparkline_requires_two_points():
    assert render_sparkline([]) == ""
    assert render_sparkline([5.0]) == ""
    svg = render_sparkline([1.0, 2.0, 3.0])
    assert svg.startswith("<svg") and "</svg>" in svg


def test_comparison_chart_normalises_and_handles_empty():
    assert render_comparison_chart({}) == ""
    assert render_comparison_chart({"A": [1.0]}) == ""
    svg = render_comparison_chart({"A": [100.0, 110.0], "B": [50.0, 45.0]})
    assert svg.count("<polyline") == 2
    assert "A" in svg and "B" in svg


def test_chart_data_uri_roundtrip():
    uri = chart_data_uri("<svg></svg>")
    assert uri.startswith("data:image/svg+xml;base64,")
    assert chart_data_uri("") == ""


def test_fallback_chart_is_a_valid_image():
    svg = fallback_chart()
    assert svg.startswith("<svg") and svg.rstrip().endswith("</svg>")


async def test_stock_chart_svg_for_known_symbol(dashboard_service):
    svg = await dashboard_service.stock_chart_svg("NOVA")
    assert svg is not None
    assert "NOVA" in svg and svg.startswith("<svg")


async def test_stock_chart_svg_is_case_insensitive(dashboard_service):
    assert await dashboard_service.stock_chart_svg("nova") is not None


async def test_stock_chart_svg_unknown_symbol_is_none(dashboard_service):
    assert await dashboard_service.stock_chart_svg("ZZZZ") is None


async def test_index_chart_svg_is_rendered(dashboard_service):
    svg = await dashboard_service.index_chart_svg()
    assert svg is not None and "Hera Exchange Index" in svg


# --------------------------------------------------------------------- the routes


async def test_overview_and_stocks_pages_are_served(client):
    for path in ("/", "/stocks"):
        response = await client.get(path)
        assert response.status == 200
        body = await response.text()
        assert "<!DOCTYPE html>" in body
        assert "/static/app.js" in body


async def test_static_assets_are_served(client):
    for path in ("/static/style.css", "/static/app.js"):
        response = await client.get(path)
        assert response.status == 200
        assert (await response.text()).strip()


async def test_market_json_endpoint(client):
    response = await client.get("/api/market")
    assert response.status == 200
    payload = await response.json()
    assert payload["stocks"]
    assert payload["index"]["value"] > 0


async def test_stocks_json_endpoint(client):
    response = await client.get("/api/stocks")
    assert response.status == 200
    payload = await response.json()
    assert len(payload["stocks"]) == 20


async def test_chart_endpoint_returns_svg(client):
    response = await client.get("/api/charts/nova.svg")
    assert response.status == 200
    assert response.content_type == "image/svg+xml"
    body = await response.text()
    assert body.startswith("<svg")
    assert "NOVA" in body


async def test_index_chart_endpoint_returns_svg(client):
    response = await client.get("/api/charts/index.svg")
    assert response.status == 200
    assert response.content_type == "image/svg+xml"
    assert (await response.text()).startswith("<svg")


async def test_chart_endpoint_404s_on_unknown_symbol(client):
    response = await client.get("/api/charts/zzzz.svg")
    assert response.status == 404


async def test_health_endpoint(client):
    response = await client.get("/healthz")
    assert response.status == 200
    payload = await response.json()
    assert payload["status"] == "ok"
    assert payload["tick"] > 0


async def test_dashboard_is_read_only(client, market):
    """Hitting every route leaves the economy untouched."""
    before = await market.snapshot(GUILD)
    for path in (
        "/",
        "/stocks",
        "/api/market",
        "/api/stocks",
        "/api/charts/nova.svg",
        "/api/charts/index.svg",
        "/healthz",
    ):
        await client.get(path)
    after = await market.snapshot(GUILD)
    assert after.tick == before.tick
    assert [c.price for c in after.companies] == [c.price for c in before.companies]


async def test_bot_starts_dashboard_once_a_guild_is_known():
    """The dashboard must wait for a guild, then bind exactly one port."""
    from hera.bot import HeraBot
    from hera.config import WebConfig, config

    original_web, original_guild = config.web, config.guild_id
    # Port 0 asks the OS for a free port, so the test never collides with a
    # real dashboard or another test run.
    object.__setattr__(config, "web", WebConfig(enabled=True, host="127.0.0.1", port=0))
    object.__setattr__(config, "guild_id", None)
    bot = HeraBot()
    try:
        await bot.db.connect()
        await bot._start_dashboard()
        assert bot.web is None, "must not bind before any guild is known"

        bot._known_guilds = {777}
        await bot._start_dashboard()
        assert bot.web is not None

        first = bot.web
        await bot._start_dashboard()
        assert bot.web is first, "on_ready firing twice must not rebind the port"
    finally:
        await bot.close()
        object.__setattr__(config, "web", original_web)
        object.__setattr__(config, "guild_id", original_guild)


async def test_bot_skips_dashboard_when_disabled(db):
    from hera.bot import HeraBot
    from hera.config import WebConfig, config

    original_web = config.web
    object.__setattr__(config, "web", WebConfig(enabled=False))
    bot = HeraBot()
    try:
        await bot.db.connect()
        bot._known_guilds = {777}
        await bot._start_dashboard()
        assert bot.web is None
    finally:
        await bot.close()
        object.__setattr__(config, "web", original_web)


def test_rendered_svg_has_no_unfilled_placeholders():
    """Guards against a format string silently slipping into the markup."""
    svg = render_price_chart("TEST", "Test Co", [(i, 10.0 + i) for i in range(10)])
    assert "{" not in svg.replace("&", "") or not re.search(r"\{[a-z_]+\}", svg)
